### PR TITLE
Make sure to enable and disable deployment logging when needed.

### PR DESCRIPTION
**Pull request by pasinskim**
Signed-off-by: Marcin Pasinski <marcin.pasinski@mender.io>

### CODE REVIEW

File: deployment_logger.go [hunk 1]
```diff
@@ -166,14 +166,86 @@ func (dlm DeploymentLogManager) Rotate() {
...
+			return file, nil
+		}
+	}
+	return "", nil
+}
+
+// GetLogs is returnig logs as a JSON string. Function is having the same
+// signature as json.Marshal() ([]byte, error)
+func (dlm DeploymentLogManager) GetLogs(deploymentID string) ([]byte, error) {
+	logFileName, err := dlm.findLogsForSpecificID(deploymentID)
```
Comment (bboozzoo) [inline deployment_logger.go]: `dlm.findLogsForSpecificID()` can return `"", nil`, I don't see `""` being handled. Maybe return `os.ErrNotExist` from `findLogsForSpecificID()`?

File: client_log.go [hunk 2]
```diff
@@ -27,16 +26,10 @@ type LogUploader interface {
...
-	Timestamp string `json:"timestamp"`
-	Level     string `json:"level"`
-	Message   string `json:"message"`
-}
-
 type LogData struct {
 	deviceID     string
 	deploymentID string
-	Messages     []LogEntry `json:"messages"`
+	Messages     []byte `json:"messages"`
```
Comment (bboozzoo) [inline client_log.go]: Any reason why we replace a structured log entry that fits what API expects with raw bytes sequence?

File: daemon_test.go [hunk 3]
```diff
@@ -121,6 +121,7 @@ func TestDaemonRun(t *testing.T) {
 		0,
 	}
 	daemon := NewDaemon(dtc, NewMemStore())
+	DeploymentLogger = NewDeploymentLogManager(".")
```
Comment (bboozzoo) [inline daemon_test.go]: Try using a `ioutil.TempDir()` instead of current directory.

File: state.go [hunk 4]
```diff
@@ -643,22 +666,24 @@ func (usr *UpdateStatusReportState) Handle(ctx *StateContext, c Controller) (Sta
...
-						Message:   "update failed",
-					},
-				})
+
+				logs, err := DeploymentLogger.GetLogs(usr.update.ID)
+				if err != nil {
+					log.Errorf("Failed to get deployment logs for deployment [%v]: %v",
+						usr.update.ID, err)
+				} else {
+					c.UploadLog(usr.update, logs)
```
Comment (bboozzoo) [inline state.go]: Either handle `UploadLog()` errors or leave a TODO note.

File: deployment_logger.go [hunk 5]
```diff
@@ -166,14 +166,86 @@ func (dlm DeploymentLogManager) Rotate() {
...
 
+	// do we have some log files already
+	if len(logFiles) == 0 {
+		return
+	}
+
 	// check if we need to delete oldest file
 	for len(logFiles) > dlm.maxLogFiles {
 		os.Remove(logFiles[0])
 		logFiles = append(logFiles[1:])
```
Comment (bboozzoo) [inline deployment_logger.go]: Can we replace this with just `logFiles = logFiles[1:]` as well?

File: client_log.go [hunk 6]
```diff
@@ -27,16 +26,10 @@ type LogUploader interface {
 	Upload(api ApiRequester, server string, logs LogData) error
 }
 
-type LogEntry struct {
-	Timestamp string `json:"timestamp"`
```
Comment (bboozzoo) [inline client_log.go]: FYI, the API will try to parse Timestamp with default format for timestamps in JSON, that happens to be RFC3339.

File: deployment_logger.go (hunk 1 above)
Comment (pasinskim) [inline deployment_logger.go]: Thanks. Will fix it!

File: client_log.go (hunk 2 above)
Comment (pasinskim) [inline client_log.go]: As agreed logs are written as a JSON entries to the log file. There is no need to marshal those and convert to `LogEntry` struct as the raw bytes returned from file are already JSON.

File: state.go (hunk 4 above)
Comment (pasinskim) [inline state.go]: Actually, maybe you have some idea how to handle this error? We are already in the error state and if we can not upload logs there is anyway nothing we can do.

File: client_log.go (hunk 6 above)
Comment (pasinskim) [inline client_log.go]: Yup. Will make sure this is the format that is written to log file.

File: deployment_logger.go (hunk 5 above)
Comment (pasinskim) [inline deployment_logger.go]: Will fix!

File: state.go (hunk 4 above)
Comment (bboozzoo) [inline state.go]: Might need to retry the upload a couple of times if it fails. I'm not sure how important it is to have these logs.

Actually, another thought, what if succeed in reporting the status but fail to upload the log, or why is this process split in two separate messages? Maybe it'd be better to report status _and_ upload logs at the same time, in a single message.

File: daemon_test.go (hunk 3 above)
Comment (pasinskim) [inline daemon_test.go]: Thanks. Fixing!

**Pull request merged**